### ISSUE TITLE
[ENHANCEMENT] handle the upgrade work with the new harvester CSI driver

### ISSUE THREAD

**Issue by Vicente-Cheng** [harvester/harvester#5294]
**Is your enhancement related to a problem? Please describe.**
The harvester CSI driver ([v0.1.4](https://github.com/harvester/harvester-csi-driver/releases/tag/v0.1.4) and later version) allow the user to specify the StorageClass they wanted. We need to add new ClusterRole and ClusterRoleBinding for the RBAC to validate it.

The creation issue is: https://github.com/harvester/harvester/issues/3467
This issue will focus on the upgrade part. 

**Describe the solution you'd like**
For v1.2.2, automatically update the ClusterRole and ClusterRoleBinding with the upgrade path.

**Describe alternatives you've considered**
None

**Additional context**
related issues: #3052

**Comment by Vicente-Cheng** [comment]
related to issue: https://github.com/kubevirt/kubevirt/issues/11677

**Comment by khushboo-rancher** [comment]
@TachunLin Could we validate it when testing 1.2.2 rc1 upgrade?

**Comment by TachunLin** [comment]
Verified **fixed** on `v1.2.2-rc1` and `v1.3-94074e85-head` with csi-driver `0.1.17`. 
Close this issue. 

### Single node Result
<details><summary>$\color{green}{\textsf{PASS}}$ Upgrade to <strong>v1.2.2-rc1</strong>, check the csi-driver ClusterRole in the RKE2 guest cluster  $~~$</summary>

* On Harvester node, we can find the ClusterRole of the csi-driver with the required content 
  ```
  harv121-upd:~ # kubectl get clusterrole harvesterhci.io:csi-driver -o yaml
  apiVersion: rbac.authorization.k8s.io/v1
  kind: ClusterRole
  metadata:
    annotations:
      meta.helm.sh/release-name: harvester
      meta.helm.sh/release-namespace: harvester-system
      objectset.rio.cattle.io/id: default-mcc-harvester-cattle-fleet-local-system
    creationTimestamp: "2024-04-25T10:30:19Z"
    labels:
      app.kubernetes.io/component: apiserver
      app.kubernetes.io/managed-by: Helm
      app.kubernetes.io/name: harvester
      app.kubernetes.io/part-of: harvester
      app.kubernetes.io/version: v1.2.2-rc1
      helm.sh/chart: harvester-1.2.2-rc1
      helm.sh/release: harvester
      objectset.rio.cattle.io/hash: e852fa897f5eae59a44b4bfe186aad80b10b94b3
    name: harvesterhci.io:csi-driver
    resourceVersion: "168760"
    uid: a9a66b2a-0308-4d7d-a0c3-35db6ad605fb
  rules:
  - apiGroups:
    - storage.k8s.io
    resources:
    - storageclasses
    verbs:
    - get
    - list
    - watch
  ```

 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Upgrade to <strong>v1.2.2-rc1</strong>, check the csi-driver ClusterRoleBinding and Service Account in the RKE2 guest cluster  $~~$</summary>

* On Harvester node, we can find the ClusterRoleBinding and ServiceAccount of the csi-driver with the required content 
  
  ```
  node1:~ # kubectl get clusterrolebinding -A | grep csi-driver
  default-default-rke2-v12712                            ClusterRole/harvesterhci.io:csi-driver                             13h
  node1:~ # kubectl get clusterrolebinding default-default-rke2-v12712 -o yaml
  apiVersion: rbac.authorization.k8s.io/v1
  kind: ClusterRoleBinding
  metadata:
    creationTimestamp: "2024-04-25T15:18:34Z"
    name: default-default-rke2-v12712
    resourceVersion: "364332"
    uid: 80ddea18-a5b7-45b3-85cc-d719b3b83fc4
  roleRef:
    apiGroup: rbac.authorization.k8s.io
    kind: ClusterRole
    name: harvesterhci.io:csi-driver
  subjects:
  - kind: ServiceAccount
    name: default-rke2-v12712
    namespace: default
  ```

 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Upgrade to <strong>v1.3.0</strong>, check the csi-driver ClusterRole in the RKE2 guest cluster  $~~$</summary>

* On Harvester node, we can find the ClusterRole of the csi-driver with the required content 
  
  ```
  harv121-upd:~ # kubectl get clusterrole harvesterhci.io:csi-driver -o yaml
  apiVersion: rbac.authorization.k8s.io/v1
  kind: ClusterRole
  metadata:
    annotations:
      meta.helm.sh/release-name: harvester
      meta.helm.sh/release-namespace: harvester-system
      objectset.rio.cattle.io/id: default-mcc-harvester-cattle-fleet-local-system
    creationTimestamp: "2024-04-25T10:30:19Z"
    labels:
      app.kubernetes.io/component: apiserver
      app.kubernetes.io/managed-by: Helm
      app.kubernetes.io/name: harvester
      app.kubernetes.io/part-of: harvester
      app.kubernetes.io/version: v1.3-d2f04e5a
      helm.sh/chart: harvester-0.0.0-v1-3-d2f04e5a
      helm.sh/release: harvester
      objectset.rio.cattle.io/hash: e852fa897f5eae59a44b4bfe186aad80b10b94b3
    name: harvesterhci.io:csi-driver
    resourceVersion: "666384"
    uid: a9a66b2a-0308-4d7d-a0c3-35db6ad605fb
  rules:
  - apiGroups:
    - storage.k8s.io
    resources:
    - storageclasses
    verbs:
    - get
    - list
    - watch
  ```


 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Upgrade to <strong>v1.3.0</strong>, check the csi-driver ClusterRoleBinding and Service Account in the RKE2 guest cluster  $~~$</summary>

* On Harvester node, we can find the ClusterRoleBinding and ServiceAccount of the csi-driver with the required content 
  
  ```
  harv121-upd:~ # kubectl get clusterrolebinding default-default-rke2-v12712 -o yaml
  apiVersion: rbac.authorization.k8s.io/v1
  kind: ClusterRoleBinding
  metadata:
    creationTimestamp: "2024-04-25T10:35:09Z"
    name: default-default-rke2-v12712
    resourceVersion: "175833"
    uid: 10f533cf-9035-4b39-a3c4-d1ce48decee0
  roleRef:
    apiGroup: rbac.authorization.k8s.io
    kind: ClusterRole
    name: harvesterhci.io:csi-driver
  subjects:
  - kind: ServiceAccount
    name: default-rke2-v12712
    namespace: default
  ```

 </details>

### Mutliple nodes Result
<details><summary>$\color{green}{\textsf{PASS}}$ Upgrade to <strong>v1.2.2-rc1</strong>, check the csi-driver ClusterRole in the RKE2 guest cluster  $~~$</summary>

* On Harvester node, we can find the ClusterRole of the csi-driver with the required content 
  
  ```
  harvester-node-1:~ # kubectl get clusterrole harvesterhci.io:csi-driver -o yaml
  apiVersion: rbac.authorization.k8s.io/v1
  kind: ClusterRole
  metadata:
    annotations:
      meta.helm.sh/release-name: harvester
      meta.helm.sh/release-namespace: harvester-system
      objectset.rio.cattle.io/id: default-mcc-harvester-cattle-fleet-local-system
    creationTimestamp: "2024-04-30T10:06:02Z"
    labels:
      app.kubernetes.io/component: apiserver
      app.kubernetes.io/managed-by: Helm
      app.kubernetes.io/name: harvester
      app.kubernetes.io/part-of: harvester
      app.kubernetes.io/version: v1.2.2-rc1
      helm.sh/chart: harvester-1.2.2-rc1
      helm.sh/release: harvester
      objectset.rio.cattle.io/hash: e852fa897f5eae59a44b4bfe186aad80b10b94b3
    name: harvesterhci.io:csi-driver
    resourceVersion: "140363"
    uid: 6baca068-0592-4d13-8d0b-bb2e89fab32b
  rules:
  - apiGroups:
    - storage.k8s.io
    resources:
    - storageclasses
    verbs:
    - get
    - list
    - watch
  ```


 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Upgrade to <strong>v1.2.2-rc1</strong>, check the csi-driver ClusterRoleBinding and Service Account in the RKE2 guest cluster  $~~$</summary>


* On Harvester node, we can find the ClusterRoleBinding and ServiceAccount of the csi-driver with the required content 
  
  ```
  harvester-node-1:~ # kubectl get clusterrolebinding -A | grep csi-driver
  default-default-rke2-v12712                            ClusterRole/harvesterhci.io:csi-driver                             165m
  harvester-node-1:~ # kubectl get clusterrolebinding default-default-rke2-v12712 -o yaml
  apiVersion: rbac.authorization.k8s.io/v1
  kind: ClusterRoleBinding
  metadata:
    creationTimestamp: "2024-04-30T10:18:31Z"
    name: default-default-rke2-v12712
    resourceVersion: "155446"
    uid: c61207f8-7220-423e-ab6f-8072766da036
  roleRef:
    apiGroup: rbac.authorization.k8s.io
    kind: ClusterRole
    name: harvesterhci.io:csi-driver
  subjects:
  - kind: ServiceAccount
    name: default-rke2-v12712
    namespace: default
  ```


 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Upgrade to <strong>v1.3.0</strong>, check the csi-driver ClusterRole in the RKE2 guest cluster  $~~$</summary>


* On Harvester node, we can find the ClusterRole of the csi-driver with the required content 
  
    ```
    harvester-node-0:~ # kubectl get clusterrole harvesterhci.io:csi-driver -o yaml
    apiVersion: rbac.authorization.k8s.io/v1
    kind: ClusterRole
    metadata:
      annotations:
        meta.helm.sh/release-name: harvester
        meta.helm.sh/release-namespace: harvester-system
        objectset.rio.cattle.io/id: default-mcc-harvester-cattle-fleet-local-system
      creationTimestamp: "2024-04-30T10:06:02Z"
      labels:
        app.kubernetes.io/component: apiserver
        app.kubernetes.io/managed-by: Helm
        app.kubernetes.io/name: harvester
        app.kubernetes.io/part-of: harvester
        app.kubernetes.io/version: v1.3-94074e85
        helm.sh/chart: harvester-0.0.0-v1-3-94074e85
        helm.sh/release: harvester
        objectset.rio.cattle.io/hash: e852fa897f5eae59a44b4bfe186aad80b10b94b3
      name: harvesterhci.io:csi-driver
      resourceVersion: "402185"
      uid: 6baca068-0592-4d13-8d0b-bb2e89fab32b
    rules:
    - apiGroups:
      - storage.k8s.io
      resources:
      - storageclasses
      verbs:
      - get
      - list
      - watch
    ```


 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Upgrade to <strong>v1.3.0</strong>, check the csi-driver ClusterRoleBinding and Service Account in the RKE2 guest cluster  $~~$</summary>


* On Harvester node, we can find the ClusterRoleBinding and ServiceAccount of the csi-driver with the required content 
  
  ```
  harvester-node-0:~ # kubectl get clusterrolebinding -A | grep csi-driver
  default-default-rke2-v12712                            ClusterRole/harvesterhci.io:csi-driver                             5h5m
  harvester-node-0:~ # kubectl get clusterrolebinding default-default-rke2-v12712 -o yaml
  apiVersion: rbac.authorization.k8s.io/v1
  kind: ClusterRoleBinding
  metadata:
    creationTimestamp: "2024-04-30T10:18:31Z"
    name: default-default-rke2-v12712
    resourceVersion: "155446"
    uid: c61207f8-7220-423e-ab6f-8072766da036
  roleRef:
    apiGroup: rbac.authorization.k8s.io
    kind: ClusterRole
    name: harvesterhci.io:csi-driver
  subjects:
  - kind: ServiceAccount
    name: default-rke2-v12712
    namespace: default
  ```



 </details>

### Test Information
* Test Environment: Three nodes baremetal machines and single node local kvm
* Harvester version: `v1.2.2-rc1` and `v1.3-94074e85-head`
* Rancher version: `v2.8.2`

### Verify Steps
<details><summary> Check the ClusterRole and ClusterRoleBinding </summary>

  1. Integrate Harvester to Rancher 
  2. Provision a RKE2 guest cluster on Harvester 
  3. Access to the Harvester node
  4. Check the content of the ClusterRole `harvesterhci.io:csi-driver` 
  correspond with https://github.com/harvester/harvester/pull/3357#issue-1528858221
  ```
  $ kubectl get clusterrole harvesterhci.io:csi-driver -o yaml
  ```
  5. Check the content of the ClusterRoleBinding and Service account of   `harvesterhci.io:csi-driver` 
  correspond with https://github.com/harvester/harvester/pull/3357#issue-1528858221
  ```
  $ kubectl get clusterrolebinding -A | grep csi-driver
  $ kubectl get clusterrolebinding <namespace>-<cluster name> -o yaml
  ```

</details>
